src/metaopt/io/parsing.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+:mod:`metaopt.io.parsing` -- Functions for configuration parsing and resolving
+==============================================================================
+
+.. module:: parsing
+   :platform: Unix
+   :synopsis: How does metaopt resolve configuration settings?
+
+ - Experiment name resolves like this:
+   cmd-arg > cmd-provided moptconfig > `DEF_CMD_EXP_NAME`
+
+ - Database options resolve with the following precedence (high to low):
+   cmd-provided moptconfig > env vars > default files > defaults
+
+.. seealso:: `ENV_VARS`, `ENV_VARS_DB`
+
+ - All other managerial, `Optimization` or `Dynamic` options resolve like this:
+   cmd-args > cmd-provided moptconfig > database (if experiment name can be
+   found) > default files
+
+Default files are given as a list at `DEF_CONFIG_FILES_PATHS` and a precedence
+is respected when building the settings dictionary:
+default metaopt example file < system-wide config < user-wide config
+
+.. note:: `Optimization` entries are required, `Dynamic` entry is optional.
+
+"""
+from __future__ import absolute_import
+import os
+import socket
+import argparse
+import textwrap
+import logging
+from copy import deepcopy
+from collections import defaultdict
+
+import six
+import yaml
+
+import metaopt
+from metaopt import (optim, dynamic)
+
+# Define type of arbitrary nested defaultdicts
+nesteddict = lambda: defaultdict(nesteddict)
+
+log = logging.getLogger(__name__)
+
+################################################################################
+#                 Default Settings and Environmental Variables                 #
+################################################################################
+
+# Default settings for command line arguments (option, description)
+DEF_CMD_MAX_TRIALS = (-1, 'inf/until preempted')
+DEF_CMD_POOL_SIZE = (10, str(10))
+DEF_CMD_EXP_NAME = (None, '{user}_{starttime}')
```

Comment:
Yeah I was wondering about that. Maybe it should be the only required argument and maybe we should enforce passing it only as a cmd arg or through cmd provided moptconfig (perhaps not even that). I do not see benefits from being able to pass it through default files.

If someone wants to resume an experiment, should explicitly type somewhere the appropriate name. Do you agree?

Reply:
That's another thing I'm not sure about. I don't think experiments should be definable *globally*. If that wasn't the case, then the user could be forced to define the experiment name in either the experiment configuration file or command-line.

I'm not sure yet what the interface would be to resume experiments but if the users needs to provide the experiment name which is by default user_starttime then that will certainly be annoying for the user to look up and provide.

Reply:
If `'exp_name'` key hits an entry in database, then we can suppose that an experiment is to be resumed (or extended if they want to use the same history of trials, but different optimization). This will work here, because database overrides defaults, but does not override cmd-provided moptconfig.

I think I will let it as it is for now and we can switch it off, if we find it not to be convenient. (I am kinda biased, in that all keyword arguments should be optional in a cli)

Reply:
I agree the majority of the options should be optional, but there are some that should not be. I see it like a git commit message. Users can write crappy commit messages, to their own misfortune, but git developers are not encouraging this by making the argument optional. Experiments name/description seems similar to me. We could even have a similar mechanism, opening an editor to write one-line name with a description below if exp_name was not given explicitly by command-line argument or from configuration files. That would be a feature for later however.